File: src/main/java/net/casualuhc/uhcmod/utils/gamesettings/GameSetting.java
package net.casualuhc.uhcmod.utils.gamesettings;

import net.casualuhc.uhcmod.UHCMod;
import net.minecraft.item.ItemStack;
import net.minecraft.nbt.NbtCompound;
import net.minecraft.nbt.NbtList;

import java.util.Map;
import java.util.Objects;
import java.util.function.Consumer;

public class GameSetting<T> {
	private final ItemStack displayStack;
	private final Map<ItemStack, T> options;
	private final Consumer<GameSetting<T>> callback;
	private T value;

	protected GameSetting(ItemStack name, Map<ItemStack, T> options, T defaultValue, Consumer<GameSetting<T>> callback) {
		this.displayStack = name;
		this.callback = callback;
		this.options = options;
		this.setValueQuietly(defaultValue);

		GameSettings.RULES.put(name, this);
	}

	public String getName() {
		return this.displayStack.getName().getString();
	}

	public ItemStack getDisplay() {
		return this.displayStack;
	}

	public Map<ItemStack, T> getOptions() {
		return this.options;
	}

	public T getValue() {
		return this.value;
	}

	public void setValueFromOption(ItemStack option) {
		T newValue = this.options.get(option);
		if (newValue != null) {
			this.setValue(newValue);
		}
	}

	public void setValue(T value) {
		this.setValueQuietly(value);
		if (this.callback != null) {
			this.callback.accept(this);
		}
	}

	public void setValueQuietly(T value) {
		if (!Objects.equals(this.value, value)) {
			this.value = value;
			this.resetSelected();
			UHCMod.LOGGER.info("Config '%s' has been set to '%s'".formatted(this.getName(), this.value));
		}
	}

	public void resetSelected() {
		this.options.forEach((stack, value) -> {
			if (this.value.equals(value)) {
				NbtCompound nbtCompound = stack.getOrCreateNbt();
				NbtList dummyList = new NbtList();
				dummyList.add(new NbtCompound());
				nbtCompound.put(ItemStack.ENCHANTMENTS_KEY, dummyList);
			} else {
				stack.removeSubNbt(ItemStack.ENCHANTMENTS_KEY);
			}
		});
	}

	public static class DoubleGameSetting extends GameSetting<Double> {
		public DoubleGameSetting(ItemStack name, Map<ItemStack, Double> options, Double defaultValue) {
			this(name, options, defaultValue, null);
		}

		public DoubleGameSetting(ItemStack name, Map<ItemStack, Double> options, Double defaultValue, Consumer<GameSetting<Double>> callback) {
			super(name, options, defaultValue, callback);
		}
	}

	public static class BooleanGameSetting extends GameSetting<Boolean> {
		public BooleanGameSetting(ItemStack name, Map<ItemStack, Boolean> options, Boolean defaultValue) {
			this(name, options, defaultValue, null);
		}

		public BooleanGameSetting(ItemStack name, Map<ItemStack, Boolean> options, Boolean defaultValue, Consumer<GameSetting<Boolean>> callback) {
			super(name, options, defaultValue, callback);
		}
	}

	public static class EnumGameSetting<E extends Enum<?>> extends GameSetting<E> {
		public EnumGameSetting(ItemStack name, Map<ItemStack, E> options, E defaultValue, Consumer<GameSetting<E>> callback) {
			super(name, options, defaultValue, callback);
		}
	}
}

File: src/main/kotlin/net/casualuhc/uhc/settings/GameSetting.kt
package net.casualuhc.uhc.utils.gamesettings

import net.casualuhc.arcade.utils.ItemUtils.enableGlint
import net.casualuhc.arcade.utils.ItemUtils.removeEnchantments
import net.casualuhc.uhc.UHCMod
import net.minecraft.world.item.ItemStack
import java.util.function.Consumer

abstract class GameSetting<T> protected constructor(
    val display: ItemStack,
    val options: Map<ItemStack, T>,
    value: T,
    private val callback: Consumer<GameSetting<T>>?
) {
    var value: T = value
        private set

    val name: String = this.display.displayName.string

    init {
        this.resetSelected()
    }

    abstract fun setFromString(string: String)

    fun setValueFromOption(option: ItemStack) {
        val newValue = options[option]
        if (newValue != null) {
            this.setValue(newValue)
        }
    }

    fun setValue(value: T) {
        setValueQuietly(value)
        if (this.callback != null) {
            this.callback.accept(this)
        }
    }

    fun setValueQuietly(value: T) {
        if (this.value != value) {
            this.value = value
            resetSelected()
            UHCMod.logger.info("Config '${this.name}' has been set to '${this.value}'")
        }
    }

    private fun resetSelected() {
        options.forEach { (stack, value) ->
            if (this.value == value) {
                stack.enableGlint()
            } else {
                stack.removeEnchantments()
            }
        }
    }

    class Int64(
        name: ItemStack,
        options: Map<ItemStack, Long>,
        defaultValue: Long,
        callback: Consumer<GameSetting<Long>>? = null
    ): GameSetting<Long>(name, options, defaultValue, callback) {
        override fun setFromString(string: String) {
            this.setValueQuietly(string.toLong())
        }
    }

    class Float64(
        name: ItemStack,
        options: Map<ItemStack, Double>,
        defaultValue: Double,
        callback: Consumer<GameSetting<Double>>? = null
    ): GameSetting<Double>(name, options, defaultValue, callback) {
        override fun setFromString(string: String) {
            this.setValueQuietly(string.toDouble())
        }
    }

    class Bool(
        name: ItemStack,
        options: Map<ItemStack, Boolean>,
        defaultValue: Boolean,
        callback: Consumer<GameSetting<Boolean>>? = null
    ): GameSetting<Boolean>(name, options, defaultValue, callback) {
        override fun setFromString(string: String) {
            this.setValueQuietly(string.toBoolean())
        }
    }

    class Enumerated<E: Enum<*>>(
        name: ItemStack,
        options: Map<ItemStack, E>,
        defaultValue: E,
        callback: Consumer<GameSetting<E>>? = null
    ): GameSetting<E>(name, options, defaultValue, callback) {
        override fun setFromString(string: String) {
            for (enum in this.options.values) {
                if (enum.name == string) {
                    this.setValueQuietly(enum)
                    return
                }
            }
            throw IllegalArgumentException("Invalid value for enum setting '$string'")
        }
    }
}
